packages/components/src/i18n/translate.tsx

```diff
@@ -0,0 +1,127 @@
+import dotobject from 'dotobject'
+import { createContext, ReactElement, useContext, useState } from 'react'
+
+export type Translation = {
+  [key: string]: string | Translation
+}
+
+export type Translations = {
+  [language: string]: Translation
+}
+type TranslateFunction = (key: string, substitutions?: any) => string
+type ConfigureFunction = ({
+  language,
+  defaultLanguage,
+  translations,
+}: {
+  language?: string
+  defaultLanguage?: string
+  translations?: Translations
+}) => void
+
+const TranslationContext = createContext(
+  {} as {
+    configure: ConfigureFunction
+    setLanguage: (lang: string) => void
+    translate: TranslateFunction
+    configuration: { language: string; defaultLanguage?: string }
+  }
+)
+
+const TranslationProvider = ({
+  language,
+  defaultLanguage,
+  translations,
+  children,
+}: {
+  language?: string
+  defaultLanguage?: string
+  translations: Translations
+  children: ReactElement
+}) => {
+  const validateConfiguration = ({
+    language,
+    defaultLanguage,
+    translations,
+  }: {
+    language?: string
+    defaultLanguage?: string
+    translations?: Translations
+  }) => {
+    if (!language && translations) {
```

Comment:
Why is better to have it like this and not using nested ifs like?

```ts
if (!translations[language]) language = language.substring(0, 2)
      if (!translations[language]) language = navigator.languages.find((language) => translations[language]) as string
           if (!translations[language])
               language = translations[defaultLanguage as string] ? defaultLanguage : Object.keys(translations)[0]

Reply:
Speaking about computation needs, yours is probably better; but we're speaking about the difference in computation of an `if(translations[language])` in case the first one has succeeded (so the next ones will be unnecessary). I think in this specific case I'd stick to the non indented version, since a single isset check won't harm (note your solution would only skip a single if, and only in one specific case: when the very first if succeeded).